code/monitor/checkmonitor.q

```diff
@@ -0,0 +1,271 @@
+
+  // and insert to checkstatus
+  `checkstatus upsert select checkid,family,metric,process,lastrun:0Np,nextrun:.z.p,status:0Nh,executiontime:0Nn,totaltime:0Nn,timerstatus:0Nh,running:0Nh,result:(count process)#() from x;  
+ }
+
+togglecheck:{[cid;status]
+  if[not cid in exec checkid from checkconfig; '"checkid ",(string cid)," doesn't exist"];
+  update active:status from `checkconfig where checkid=cid;
+ }
+
+disablecheck:togglecheck[;0b]
+enablecheck:togglecheck[;1b]
+
+// input to runcheck will be a row from checkconfig
+runcheck:{
+  // increment the run id
+  runid+:1i;
+  // get the handle to the process
+ if[null h:gethandle[x`process]; `checkstatus upsert ((enlist`checkid)!enlist x`checkid),update running:0h,result:"no handle connection",status:0h,timerstatus:0h from checkstatus x`checkid];
+  // run the check remotely
+  // send over the id, return a dict of `id`status`res
+  .async.postback[h;({start:.z.p; (`runid`executiontime!(y;.z.p-start)),`status`result!@[{(1h;value x)};x;{(0h;x)}]};(x`query;x`params);runid);`checkresulthandler];
+
+  // add a record to track 
+  `checktracker insert (runid;.z.p;0Np;0Nn;x`checkid;0Nh;());
+  // update the status to be running 
+  `checkstatus upsert ((enlist`checkid)!enlist x`checkid),update running:1h from checkstatus x`checkid;
+ } 
+
+//check that process has not been running over next allotted runtime
+//if so, set status and timerstatus to neg
+checkruntime:{[n]
+  update status:0h,timerstatus:0h,running:0h from `checkstatus where running=1h,n<.z.p-nextrun;
+ }
+
+checkresulthandler:{
+  // update the appropriate record in checktracker
+  toinsert:((enlist`runid)#x),(checktracker x`runid),((enlist`receivetime)!enlist .z.p),`executiontime`status`result#x;
+
+  // store the record
+  `checktracker upsert toinsert; 
+
+  // get the configuration for this alert 
+  conf:checkconfig toinsert`checkid;
+ 
+  // need to run the resultchecker against the actual result
+  // but only if the actual query has been successful
+   if[x`status; 
+   // pull out the resultchecker function and run it
+   // this should only get triggered in dev cycle- 
+   // protect against devs inserting incorrect resultchecker definitions
+   r:@[value;(conf`resultchecker;conf`params;toinsert);{`status`result!(0b;"failed to run result checker for runid ",(string x`runid)," and checkid ",(string x`checkid),": ",y)}[toinsert]];
+   // have to make sure we have dictionary result
+  if[not 99h=type r; r:`status`result!(0b;"resultchecker function did not return a dictionary")];
+   
+  // check here if it has failed or passed
+  // override the status and error message as appropriate
+  toinsert[`status]:toinsert[`status] and r`status;
+  toinsert[`result]:r`result; 
+  ];
+ 
+ // if the query has failed, add in the error
+  if[not x`status; toinsert[`result]:"request failed on remote server: ",x`result];
+ 
+  // insert the record into checkstatus
+  `checkstatus upsert ((enlist `checkid)!enlist toinsert`checkid),
```

Comment:
Might be the same as either of these:
```
enlist[`checkid]#toinsert;
(1#`checkid)#toinsert;
```

Reply:
just tried these, they don't seem to upsert the time data